persistent-postgresql/Database/Persist/Postgresql.hs

```diff
@@ -340,71 +335,14 @@ withStmt' :: MonadIO m
-      rawquery <- PG.formatQuery conn query (map P vals)
-
-      -- Take raw connection
-      (rt, rr, rc, ids) <- PG.withConnection conn $ \rawconn -> do
-            -- Execute query
-            mret <- LibPQ.exec rawconn rawquery
-            case mret of
-              Nothing -> do
-                merr <- LibPQ.errorMessage rawconn
-                fail $ case merr of
-                         Nothing -> "Postgresql.withStmt': unknown error"
-                         Just e  -> "Postgresql.withStmt': " ++ B8.unpack e
-              Just ret -> do
-                -- Check result status
-                status <- LibPQ.resultStatus ret
-                case status of
-                  LibPQ.TuplesOk -> return ()
-                  _ -> PG.throwResultError "Postgresql.withStmt': bad result status " ret status
-
-                -- Get number and type of columns
-                cols <- LibPQ.nfields ret
-                oids <- forM [0..cols-1] $ \col -> fmap ((,) col) (LibPQ.ftype ret col)
-                -- Ready to go!
-                rowRef   <- newIORef (LibPQ.Row 0)
-                rowCount <- LibPQ.ntuples ret
-                return (ret, rowRef, rowCount, oids)
-      let getters
-            = map (\(col, oid) -> getGetter conn oid $ PG.Field rt col oid) ids
-      return (rt, rr, rc, getters)
-
-    closeS (ret, _, _, _) = LibPQ.unsafeFreeResult ret
-
-    pull x = do
-        y <- liftIO $ pullS x
-        case y of
-            Nothing -> return ()
-            Just z -> yield z >> pull x
-
-    pullS (ret, rowRef, rowCount, getters) = do
-        row <- atomicModifyIORef rowRef (\r -> (r+1, r))
-        if row == rowCount
-           then return Nothing
-           else fmap Just $ forM (zip getters [0..]) $ \(getter, col) -> do
-                                mbs <- LibPQ.getvalue' ret row col
-                                case mbs of
-                                  Nothing ->
-                                    -- getvalue' verified that the value is NULL.
-                                    -- However, that does not mean that there are
-                                    -- no NULL values inside the value (e.g., if
-                                    -- we're dealing with an array of optional values).
-                                    return PersistNull
-                                  Just bs -> do
-                                    ok <- PGFF.runConversion (getter mbs) conn
-                                    bs `seq` case ok of
-                                                        Errors (exc:_) -> throw exc
-                                                        Errors [] -> error "Got an Errors, but no exceptions"
-                                                        Ok v  -> return v
+    mkAcquire openS closeS
+    where
+        closeS _= return ()
+        openS  = PG.fold conn query (map P vals) CL.sourceNull processRow
+        processRow s row = return $ s >> yield (map pVal row)
```

Comment:
I really like how this simplifies the code.

Docs for [`fold`](https://www.stackage.org/haddock/lts-14.19/postgresql-simple-0.6.2/Database-PostgreSQL-Simple.html#v:fold).

> This fold is not strict. The stream consumer is responsible for forcing the evaluation of its result to avoid space leaks.

We may want to provide a bang pattern on the `row`.

Using `return ()` is fine with `PG.fold` as a finalizer, because `fold` itself calls `bracket` and cleans up it's own resources. There may be another issue here.

Reply:
Fixed in https://github.com/yesodweb/persistent/pull/726/commits/7e39b02095e39560bfa373f1c6cb3da1e7752211